### ISSUE TITLE
[Question] better way to reference outlet inside UITableViewCell?

### ISSUE THREAD

**Issue by hermanccw** [schibsted/layout#68]
Hi there,

I have a cell within a table defined in XML like this:
```
<UITableViewCell reuseIdentifier="cell">
            <UITextField
                autocapitalizationType="words"
                height="100%"
                isSecureTextEntry="isPassword ? true : false"
                left="15"
                placeholder="{placeholder}"
                text="{value}"
                textAlignment="left"
                width="100%"
            />
</UITableViewCell>
```

I wanted to reference the `UITextField` inside the cell in my view controller so that whenever text chnaged within a textField, I get notify and do something about it.  Here is my current solution to get the reference and add target to the text field:
```
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let node = tableView.dequeueReusableCellNode(withIdentifier: "cell", for: indexPath)
    
    node.setState(viewModel.formValues[indexPath.row])
    
    let textField = node.children.first?.view as! UITextField
    textField.tag = indexPath.row
    textField.removeTarget(nil, action: nil, for: .allEvents)
    textField.addTarget(self, action: #selector(SignUpFormViewController.textFieldDidChanged(textField:)), for: .editingChanged)

    return node.view as! UITableViewCell
  }
```

Is there a more elegant way to reference outlet without a cell?

**Comment by nicklockwood** [comment]
The cleanest approach is to create a subclass of UITableViewCell, with an outlet for the textField. Then you can bind the textField to the outlet in xml and access it directly from the cell.

**Comment by comm1x** [comment]
I am doing like Nick said - creating subclass and incapsulating all logic into it. There is a short sample:

View layout:
```
<UITableViewCell
    width="parent.width"
    height="0">

    <!--Hall Name-->
    <UILabel
        outlet="hallName"
        top="6"
        left="50"
        right="parent.width - 50" />

    <!--City Name-->
    <UILabel
        outlet="cityName"
        top="30"
        left="50"
        right="parent.width - 50"
        font="15"
        textColor="darkGray" />

</UITableViewCell>
```

Controller of cell:
```
class HallView: UITableViewCell, LayoutLoading {
    @IBOutlet var hallName: UILabel!
    @IBOutlet var cityName: UILabel!

    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        loadLayout(named: "HallView.xml")
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        self.layoutNode?.update()
    }

    func loadModel(hall: Hall) {
        hallName.text = hall.name
        cityName.text = hall.city
    }
}
```

Use into screen layout:
```
<UIViewController>
    <UITableView
        outlet="tableView">
            <HallView reuseIdentifier="HallView"/>
    </UITableView>
</UIViewController>
```

And into screen controller:
```
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let node = tableView.dequeueReusableCellNode(withIdentifier: "HallView", for: indexPath)
        let cell = node.view as! HallView
        cell.loadModel(hall: halls[indexPath.row])
        return cell
    }
```

**Comment by silentninja** [comment]
I am trying to achieve the same by subclassing UITableViewCell and using LayoutLoading but I am getting an error saying it is not a Layout managed view
https://github.com/schibsted/layout/blob/a410461ab8ef29f2cc44ae2c3eb1d11b011ee84d/Layout/UITableView%2BLayout.swift#L407
```
preconditionFailure("\(type(of: cell)) is not a Layout-managed view")
```
This happens when I register the cell class instead of the layout
```
 tableView?.register(ExampleCell.self, forCellReuseIdentifier: "exampleCell")
```

When I create the cell based on xml, I am getting the following error 
```
Could not cast value of type 'NSKVONotifying_UITableViewCell' (0x6040001122d0) to 'SampleProject.ExampleCell'
```
My UITableViewCell class 
```
import UIKit
import Layout
class ExampleCell: UITableViewCell, LayoutLoading {


    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        loadLayout(named: "example_cell.xml")
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        self.layoutNode?.update()
    }

}
```
My TableViewController class
```
class SampleViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

    @IBOutlet var tableView: UITableView?


    func tableView(_: UITableView, numberOfRowsInSection _: Int) -> Int {
        return 2
    }


    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let identifier = "exampleCell"
        let node = tableView.dequeueReusableCellNode(withIdentifier: identifier, for: indexPath)
        let cell = node.view as! ExampleCell
        return cell
    }
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }



}
```
Xml layout for TableView:
```
    <SampleViewController
            title="Explore"
            backgroundColor="white"
            >
        <UITableView
                backgroundColor="tintColor"
                contentInset.bottom="safeAreaInsets.bottom"
                contentInset.top="safeAreaInsets.top"
                contentInsetAdjustmentBehavior="never"
                contentOffset.y="-safeAreaInsets.top"
                outlet="tableView"
                scrollIndicatorInsets.bottom="safeAreaInsets.bottom"
                scrollIndicatorInsets.top="safeAreaInsets.top"
                style="plain"/>
        <SampleCell reuseIdentifier="sampleCell"/>

    </SampleViewController>
```
Xml for TableViewCell:
```
<UIView
        width="parent.width"
        height="0">

    <!--Hall Name-->
    <UILabel
            top="6"
            left="50"
            right="parent.width - 50" />

    <!--City Name-->
    <UILabel
            top="30"
            left="50"
            right="parent.width - 50"
            font="15"
            textColor="darkGray" />

</UIView>
```

**Comment by comm1x** [comment]
@silentninja I am not sure, but I notice, you use different identifier: 

```
let identifier = "exampleCell"
<SampleCell reuseIdentifier="sampleCell"/>
```

**Comment by silentninja** [comment]
Sorry for the confusion, made a mistake when renaming, fixed the code now. I am using correct identifier in my code though and problem still happens

**Comment by comm1x** [comment]
@silentninja 
```
<UITableView
                backgroundColor="tintColor"
                contentInset.bottom="safeAreaInsets.bottom"
                contentInset.top="safeAreaInsets.top"
                contentInsetAdjustmentBehavior="never"
                contentOffset.y="-safeAreaInsets.top"
                outlet="tableView"
                scrollIndicatorInsets.bottom="safeAreaInsets.bottom"
                scrollIndicatorInsets.top="safeAreaInsets.top"
                style="plain"/>
        <SampleCell reuseIdentifier="exampleCell"/>
```
Is it correct? You declare SampleCell outside of TableView. It should be inside.

**Comment by silentninja** [comment]
That seems to be the problem thanks for correcting it, I am able to get it working now. I am facing problem with
 1. Registering the cell as a class 
2. The cell taking the whole screen when the height is set to auto. This one is probably some mistake on my end. Will

**Comment by comm1x** [comment]
@silentninja cell automacally registered, when you put cell-class inside tableview-class in xml files.
```
<UITableView outlet="tableView">
        <HallView reuseIdentifier="HallView"/>
</UITableView>
```

With automatic height I am facing too. Write about could you solve this problem.

**Comment by silentninja** [comment]
I am facing problem with registering the cell as a class in swift 
 tableView?.register(ExampleCell.self, forCellReuseIdentifier: "exampleCell")``` 

```
 instead of adding it in xml
```
<UITableView outlet="tableView">
        <ExampleCell reuseIdentifier="exampleCell"/>
</UITableView>
```
As for the automatic height I had to define it explicitly with height="auto" and width="100%" for the width. The top,left,right of root node of the cell doesn't work well(content crosses the cell), I ended up wrapping it inside a UIView whose height is set to auto and everything seems to work fine

**Comment by nicklockwood** [comment]
@comm1x I can see why this would be confusing, but there is no need for your custom ExampleCell to implement the `LayoutLoading` protocol, and I suspect that's why you are having issues.

`LayoutLoading` is for views that load a layout internally, but are not themselves part of a Layout hierarchy. Loading of UITableViewCell layouts is already handled by the tableview itself, when you us the `registerLayout(named:forCellReuseIdentifier:)` method, or when you nest the xml for your cell inside the layout xml for the table.

By making the custom cell conform to the `LayoutLoading` protocol, and load its own content, you are loading the cell twice, creating a cell within a cell.

**Comment by nicklockwood** [comment]
I'll try to make this clearer in the documentation.

**Comment by silentninja** [comment]
@nicklockwood I tried the nesting the views under the UITableViewCell subclass but I couldn't get access to the layoutNode, so I had to use  the layoutloading protocol.
 ```
//when accessing layoutNode
self.layoutNode?.setState()

Error:(28, 20) 'layoutNode' is inaccessible due to 'fileprivate' protection level
```
If I dont use self I get the layoutNode as nil

I need to access layoutNode to update the state of the data when some updates happen to the data when the event is triggered.


Nesting the cells instead of using layoutLoading protocol
```
// example_cell.xml
<SampleCell reuseIdentifier="exampleCell">
<UIView
        width="parent.width"
        height="0">

    <!--Hall Name-->
    <UILabel
            top="6"
            left="50"
            right="parent.width - 50" />

    <!--City Name-->
    <UILabel
            top="30"
            left="50"
            right="parent.width - 50"
            font="15"
            textColor="darkGray" />

</UIView>
</SampleCell>
```

**Comment by silentninja** [comment]
I have created a new issue for the previous comment here #79.

**Comment by nicklockwood** [comment]
@hermanccw I've made the `layoutNode` property of `UITableViewCell` public, and also added a `childNode(withID:)` method to `LayoutNode` as an alternative solution for the original problem of referencing sub-nodes of a cell.

If you link against the develop branch by putting this in your Podfile, you can use these features:

pod 'Layout', :git => 'https://github.com/schibsted/Layout.git', :branch => 'develop'

**Comment by nicklockwood** [comment]
Fixed in 0.6.0
